File: SpaceVIL/Decorations/Effects.cs
using System.Collections.Generic;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal class Effects : IAppearanceExtension
    {
        private HashSet<IEffect> _subtract = new HashSet<IEffect>();
        private List<IEffect> _shadow = new List<IEffect>();
        private List<IEffect> _border = new List<IEffect>();

        internal Effects() { }

        public void Add(IEffect effect)
        {
            if (effect is ISubtractFigure)
            {
                _subtract.Add(effect);
                return;
            }
            else if (effect is IShadow)
            {
                _shadow.Add(effect);
                return;
            }
            else if (effect is IBorder)
            {
                _border.Add(effect);
                return;
            }
        }

        public void Remove(IEffect effect)
        {
            if (effect is ISubtractFigure)
            {
                _subtract.Remove(effect);
                return;
            }
            else if (effect is IShadow)
            {
                _shadow.Remove(effect);
                return;
            }
            else if (effect is IBorder)
            {
                _border.Remove(effect);
                return;
            }
        }

        public List<IEffect> Get(EffectType type)
        {
            switch (type)
            {
                case EffectType.Border:
                    return new List<IEffect>(_border);
                case EffectType.Shadow:
                    return new List<IEffect>(_shadow);
                case EffectType.Subtract:
                    return new List<IEffect>(_subtract);
            }
            return new List<IEffect>();
        }

        public void Clear(EffectType type)
        {
            switch (type)
            {
                case EffectType.Border:
                    _border.Clear();
                    break;
                case EffectType.Shadow:
                    _shadow.Clear();
                    break;
                case EffectType.Subtract:
                    _subtract.Clear();
                    break;
            }
        }

        public void Clear()
        {
            _border.Clear();
            _shadow.Clear();
            _subtract.Clear();
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/Effects.java
package com.spvessel.spacevil;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import com.spvessel.spacevil.Core.IAppearanceExtension;
import com.spvessel.spacevil.Core.IBorder;
import com.spvessel.spacevil.Core.IEffect;
import com.spvessel.spacevil.Core.IShadow;
import com.spvessel.spacevil.Core.ISubtractFigure;
import com.spvessel.spacevil.Flags.EffectType;

class Effects implements IAppearanceExtension {
    private Set<IEffect> _subtract = new HashSet<>();
    private ArrayList<IEffect> _shadow = new ArrayList<>();
    private ArrayList<IEffect> _border = new ArrayList<>();

    Effects() {
    }

    @Override
    public void add(IEffect effect) {
        if (effect instanceof ISubtractFigure) {
            _subtract.add(effect);
            return;
        } else if (effect instanceof IShadow) {
            _shadow.add(effect);
            return;
        } else if (effect instanceof IBorder) {
            _border.add(effect);
            return;
        }
    }

    @Override
    public void remove(IEffect effect) {
        if (effect instanceof ISubtractFigure) {
            _subtract.remove(effect);
            return;
        } else if (effect instanceof IShadow) {
            _shadow.remove(effect);
            return;
        } else if (effect instanceof IBorder) {
            _border.remove(effect);
            return;
        }
    }

    @Override
    public List<IEffect> get(EffectType type) {
        switch (type) {
            case Border:
                return new ArrayList<>(_border);
            case Shadow:
                return new ArrayList<>(_shadow);
            case Subtract:
                return new ArrayList<>(_subtract);
        }
        return new ArrayList<IEffect>();
    }

    @Override
    public void clear(EffectType type) {
        switch (type) {
            case Border:
                _border.clear();
                _border.trimToSize();
                break;
            case Shadow:
                _shadow.clear();
                _shadow.trimToSize();
                break;
            case Subtract:
                _subtract.clear();
                break;
        }
    }

    @Override
    public void clear() {
        _border.clear();
        _shadow.clear();
        _subtract.clear();
        _border.trimToSize();
        _shadow.trimToSize();
    }
}
